### PR TITLE
Enable download of future forks

### PR DESCRIPTION
**Pull request by arkpar**
Currently sync won't download forks ahead of the best chain announced with `is_best=false` This PR removes this limitation as long as the fork is not too long.

### CODE REVIEW

File: client/network/src/protocol/sync.rs [hunk 1]
```diff
@@ -2418,7 +2418,11 @@ fn fork_sync_request<B: BlockT>(
 		if !r.peers.contains(id) {
 			continue
 		}
-		if r.number <= best_num {
+		// Download the fork only if it is behind or not too far ahead our tip of the chain
+		// Otherwise it should be downloaded in full sync mode.
+		if r.number <= best_num ||
+			(r.number - best_num).saturated_into::<u32>() < MAX_BLOCKS_TO_REQUEST as u32
```
Comment (rphmeier) [inline client/network/src/protocol/sync.rs:2424]: @eskimor let's hope disputes travel & resolve quickly! Haha

Seriously, though, this seems brittle so kind of OK for a dirty patch but we should have a think about doing this 'right' which will involve changes outside of sync about fork-choice.

**Comment by rphmeier** [comment]
I'm waiting on giving an approval as a result of testing on Versi to make sure this alleviates the stalls we've been seeing.

File: client/network/test/src/sync.rs [hunk 2]
```diff
@@ -847,7 +886,10 @@ fn block_announce_data_is_propagated() {
 	// Wait until peer 1 is connected to both nodes.
 	block_on(futures::future::poll_fn::<(), _>(|cx| {
 		net.poll(cx);
-		if net.peer(1).num_peers() == 2 {
+		if net.peer(1).num_peers() == 2 &&
+			net.peer(0).num_peers() == 1 &&
+			net.peer(2).num_peers() == 1
```
Comment (bkchr) [inline client/network/test/src/sync.rs:891]: This should be the consequence of  `net.peer(1).num_peers() == 2`?

Comment (bkchr) [reply client/network/test/src/sync.rs:891]: Aka we don't need it?

**Review by bkchr — APPROVED** [review]

**Comment by arkpar** [comment]
Is this still being tested?

**Comment by bkchr** [comment]
> Is this still being tested?

AFAIK it was successful, let us merge it.

**Comment by bkchr** [comment]
bot merge

**Pull request merged**